In [ ]:
"""Statistical Analysis with Pandas"""

'Statistical Analysis with Pandas'

In [ ]:
#!usr/bin/env/ python

# Lesson 3 Assignment

In [ ]:
import pandas as pd

1. Import data mtcars.csv. Download mtcars.csv into Python. (10 points)

In [ ]:
cars = pd.read_csv('mtcars.csv')

2. Explore the data and perform a statistical analysis of the data. (30 points)

We begin our exploration by simply viewing the dataframe. A short description of each field can be seen below.

![image.png](attachment:image.png)

In [ ]:
cars

model   mpg  cyl   disp   hp  drat     wt   qsec  vs  am  \
0             Mazda RX4  21.0    6  160.0  110  3.90  2.620  16.46   0   1   
1         Mazda RX4 Wag  21.0    6  160.0  110  3.90  2.875  17.02   0   1   
2            Datsun 710  22.8    4  108.0   93  3.85  2.320  18.61   1   1   
3        Hornet 4 Drive  21.4    6  258.0  110  3.08  3.215  19.44   1   0   
4     Hornet Sportabout  18.7    8  360.0  175  3.15  3.440  17.02   0   0   
5               Valiant  18.1    6  225.0  105  2.76  3.460  20.22   1   0   
6            Duster 360  14.3    8  360.0  245  3.21  3.570  15.84   0   0   
7             Merc 240D  24.4    4  146.7   62  3.69  3.190  20.00   1   0   
8              Merc 230  22.8    4  140.8   95  3.92  3.150  22.90   1   0   
9              Merc 280  19.2    6  167.6  123  3.92  3.440  18.30   1   0   
10            Merc 280C  17.8    6  167.6  123  3.92  3.440  18.90   1   0   
11           Merc 450SE  16.4    8  275.8  180  3.07  4.070  17.40   0   0   
12           Merc 450SL  17.3    8  275.8  180  3.07  3.730  17.60   0   0   
13          Merc 450SLC  15.2    8  275.8  180  3.07  3.780  18.00   0   0   
14   Cadillac Fleetwood  10.4    8  472.0  205  2.93  5.250  17.98   0   0   
15  Lincoln Continental  10.4    8  460.0  215  3.00  5.424  17.82   0   0   
16    Chrysler Imperial  14.7    8  440.0  230  3.23  5.345  17.42   0   0   
17             Fiat 128  32.4    4   78.7   66  4.08  2.200  19.47   1   1   
18          Honda Civic  30.4    4   75.7   52  4.93  1.615  18.52   1   1   
19       Toyota Corolla  33.9    4   71.1   65  4.22  1.835  19.90   1   1   
20        Toyota Corona  21.5    4  120.1   97  3.70  2.465  20.01   1   0   
21     Dodge Challenger  15.5    8  318.0  150  2.76  3.520  16.87   0   0   
22          AMC Javelin  15.2    8  304.0  150  3.15  3.435  17.30   0   0   
23           Camaro Z28  13.3    8  350.0  245  3.73  3.840  15.41   0   0   
24     Pontiac Firebird  19.2    8  400.0  175  3.08  3.845  17.05   0   0   
25            Fiat X1-9  27.3    4   79.0   66  4.08  1.935  18.90   1   1   
26        Porsche 914-2  26.0    4  120.3   91  4.43  2.140  16.70   0   1   
27         Lotus Europa  30.4    4   95.1  113  3.77  1.513  16.90   1   1   
28       Ford Pantera L  15.8    8  351.0  264  4.22  3.170  14.50   0   1   
29         Ferrari Dino  19.7    6  145.0  175  3.62  2.770  15.50   0   1   
30        Maserati Bora  15.0    8  301.0  335  3.54  3.570  14.60   0   1   
31           Volvo 142E  21.4    4  121.0  109  4.11  2.780  18.60   1   1   

    gear  carb  
0      4     4  
1      4     4  
2      4     1  
3      3     1  
4      3     2  
5      3     1  
6      3     4  
7      4     2  
8      4     2  
9      4     4  
10     4     4  
11     3     3  
12     3     3  
13     3     3  
14     3     4  
15     3     4  
16     3     4  
17     4     1  
18     4     2  
19     4     1  
20     3     1  
21     3     2  
22     3     2  
23     3     4  
24     3     2  
25     4     1  
26     5     2  
27     5     2  
28     5     4  
29     5     6  
30     5     8  
31     4     2

There are 32 rows of data and 12 columns. The `model` column contains nominal categorical data. The columns `vs` and `am` are both ordinal categorical data. The columns `vs` and  `am` are binary. All other columns shown are numerical. `cyl`, `gear`, and `carb` are discrete columns and all others contain continous numerical data. Each row of data represents a unique model and has various information about the car.

Next, I apply the `describe()` method  to see various descriptive statistics for all quantitative columns of the dataframe.

In [ ]:
cars.drop(['vs', 'am'], axis=1).describe()

mpg        cyl        disp          hp       drat         wt  \
count  32.000000  32.000000   32.000000   32.000000  32.000000  32.000000   
mean   20.090625   6.187500  230.721875  146.687500   3.596563   3.217250   
std     6.026948   1.785922  123.938694   68.562868   0.534679   0.978457   
min    10.400000   4.000000   71.100000   52.000000   2.760000   1.513000   
25%    15.425000   4.000000  120.825000   96.500000   3.080000   2.581250   
50%    19.200000   6.000000  196.300000  123.000000   3.695000   3.325000   
75%    22.800000   8.000000  326.000000  180.000000   3.920000   3.610000   
max    33.900000   8.000000  472.000000  335.000000   4.930000   5.424000   

            qsec       gear     carb  
count  32.000000  32.000000  32.0000  
mean   17.848750   3.687500   2.8125  
std     1.786943   0.737804   1.6152  
min    14.500000   3.000000   1.0000  
25%    16.892500   3.000000   2.0000  
50%    17.710000   4.000000   2.0000  
75%    18.900000   4.000000   4.0000  
max    22.900000   5.000000   8.0000

Then, I begin to explore the qualitative columns: `model`, `vs`, and `am`. Applying the `unique` method to the `model` column allows me to see all distinct car models represented by the rows.

In [ ]:
cars.model.unique

<bound method Series.unique of 0               Mazda RX4
1           Mazda RX4 Wag
2              Datsun 710
3          Hornet 4 Drive
4       Hornet Sportabout
5                 Valiant
6              Duster 360
7               Merc 240D
8                Merc 230
9                Merc 280
10              Merc 280C
11             Merc 450SE
12             Merc 450SL
13            Merc 450SLC
14     Cadillac Fleetwood
15    Lincoln Continental
16      Chrysler Imperial
17               Fiat 128
18            Honda Civic
19         Toyota Corolla
20          Toyota Corona
21       Dodge Challenger
22            AMC Javelin
23             Camaro Z28
24       Pontiac Firebird
25              Fiat X1-9
26          Porsche 914-2
27           Lotus Europa
28         Ford Pantera L
29           Ferrari Dino
30          Maserati Bora
31             Volvo 142E
Name: model, dtype: object>

Next, I apply the `value_counts()` method to the `am` column to see how many cars have automatic versus manual transmissions. I apply the same method to `vs` to see how have cars have a V engine versus a straight engine. There appears to be 19 automatic cars and 13 manual cars. Also, 18 of the cars have V engines while 14 have straight engines.

In [ ]:
cars.am.value_counts()

am
0    19
1    13
Name: count, dtype: int64

In [ ]:
cars.vs.value_counts()

vs
0    18
1    14
Name: count, dtype: int64

Finally, I use the `corr()` method to explore the relationship between all qualitative attributes. `mpg` appears to have many strong positive and negative relationships with many attributes. The strongest relationship between all columns are `cyl` and `disp` with a correlation of 0.902.

In [ ]:
cars.drop(['model', 'vs', 'am'], axis=1).corr()

mpg       cyl      disp        hp      drat        wt      qsec  \
mpg   1.000000 -0.852162 -0.847551 -0.776168  0.681172 -0.867659  0.418684   
cyl  -0.852162  1.000000  0.902033  0.832447 -0.699938  0.782496 -0.591242   
disp -0.847551  0.902033  1.000000  0.790949 -0.710214  0.887980 -0.433698   
hp   -0.776168  0.832447  0.790949  1.000000 -0.448759  0.658748 -0.708223   
drat  0.681172 -0.699938 -0.710214 -0.448759  1.000000 -0.712441  0.091205   
wt   -0.867659  0.782496  0.887980  0.658748 -0.712441  1.000000 -0.174716   
qsec  0.418684 -0.591242 -0.433698 -0.708223  0.091205 -0.174716  1.000000   
gear  0.480285 -0.492687 -0.555569 -0.125704  0.699610 -0.583287 -0.212682   
carb -0.550925  0.526988  0.394977  0.749812 -0.090790  0.427606 -0.656249   

          gear      carb  
mpg   0.480285 -0.550925  
cyl  -0.492687  0.526988  
disp -0.555569  0.394977  
hp   -0.125704  0.749812  
drat  0.699610 -0.090790  
wt   -0.583287  0.427606  
qsec -0.212682 -0.656249  
gear  1.000000  0.274073  
carb  0.274073  1.000000

## 3. Analyze mpg for cars with different gear, and show your findings. (20 points)

In [ ]:
cars.groupby("gear")['mpg'].describe()

count       mean       std   min   25%   50%     75%   max
gear                                                            
3      15.0  16.106667  3.371618  10.4  14.5  15.5  18.400  21.5
4      12.0  24.533333  5.276764  17.8  21.0  22.8  28.075  33.9
5       5.0  21.380000  6.658979  15.0  15.8  19.7  26.000  30.4

Miles per gallon is greatest when gear is 4 and least when gear is 3.

## 4. Analyze mpg for cars with different carb, and show your findings. (20 points)

In [ ]:
cars.groupby("carb")['mpg'].describe()

count       mean       std   min     25%    50%    75%   max
carb                                                              
1       7.0  25.342857  6.001349  18.1  21.450  22.80  29.85  33.9
2      10.0  22.400000  5.472152  15.2  18.825  22.10  25.60  30.4
3       3.0  16.300000  1.053565  15.2  15.800  16.40  16.85  17.3
4      10.0  15.790000  3.911081  10.4  13.550  15.25  18.85  21.0
6       1.0  19.700000       NaN  19.7  19.700  19.70  19.70  19.7
8       1.0  15.000000       NaN  15.0  15.000  15.00  15.00  15.0

Miles per gallon is greatest when carb is 1 and least when carb is 8. MPG varies greatly when carb is 1, with the minimum mpg being 18.1 and the maximum being 33.9.

## 5. Find out which attribute has the most impact on mpg. (20 points)

In [ ]:
cars.drop(['model', 'vs', 'am'], axis=1).corrwith(cars.mpg)

mpg     1.000000
cyl    -0.852162
disp   -0.847551
hp     -0.776168
drat    0.681172
wt     -0.867659
qsec    0.418684
gear    0.480285
carb   -0.550925
dtype: float64

In [ ]:
cars.drop('model', axis=1).cov()

mpg         cyl          disp           hp       drat  \
mpg    36.324103   -9.172379   -633.097208  -320.732056   2.195064   
cyl    -9.172379    3.189516    199.660282   101.931452  -0.668367   
disp -633.097208  199.660282  15360.799829  6721.158669 -47.064019   
hp   -320.732056  101.931452   6721.158669  4700.866935 -16.451109   
drat    2.195064   -0.668367    -47.064019   -16.451109   0.285881   
wt     -5.116685    1.367371    107.684204    44.192661  -0.372721   
qsec    4.509149   -1.886855    -96.051681   -86.770081   0.087141   
vs      2.017137   -0.729839    -44.377621   -24.987903   0.118649   
am      1.803931   -0.465726    -36.564012    -8.320565   0.190151   
gear    2.135685   -0.649194    -50.802621    -6.358871   0.275988   
carb   -5.363105    1.520161     79.068750    83.036290  -0.078407   

              wt       qsec         vs         am       gear       carb  
mpg    -5.116685   4.509149   2.017137   1.803931   2.135685  -5.363105  
cyl     1.367371  -1.886855  -0.729839  -0.465726  -0.649194   1.520161  
disp  107.684204 -96.051681 -44.377621 -36.564012 -50.802621  79.068750  
hp     44.192661 -86.770081 -24.987903  -8.320565  -6.358871  83.036290  
drat   -0.372721   0.087141   0.118649   0.190151   0.275988  -0.078407  
wt      0.957379  -0.305482  -0.273661  -0.338105  -0.421081   0.675790  
qsec   -0.305482   3.193166   0.670565  -0.204960  -0.280403  -1.894113  
vs     -0.273661   0.670565   0.254032   0.042339   0.076613  -0.463710  
am     -0.338105  -0.204960   0.042339   0.248992   0.292339   0.046371  
gear   -0.421081  -0.280403   0.076613   0.292339   0.544355   0.326613  
carb    0.675790  -1.894113  -0.463710   0.046371   0.326613   2.608871

Although disp has the third strongest negative correlation with mps behind wt and cyl, disp has the largest negative covariance by far, indicating a strong negative relationship between disp and mph. Because of this, I can assume disp has the most impact on mpg